src/app/fdctl/run/tiles/fd_repair.c

```diff
@@ -0,0 +1,675 @@
+                            ulong buf_sz ) {
+
+  fd_repair_request_t const * repair_reqs = (fd_repair_request_t const *)fd_type_pun_const( buf );
+  ulong repair_req_cnt = buf_sz;
+  // FD_LOG_WARNING(("Repair requests %lu received %lu", repair_req_cnt, buf_sz));
+  for( ulong i = 0; i < repair_req_cnt; i++ ) {
+    fd_repair_request_t const * repair_req = &repair_reqs[i];
+    int rc = 0;
+    switch(repair_req->type) {
+      case FD_REPAIR_REQ_TYPE_NEED_WINDOW_INDEX: {
+        rc = fd_repair_need_window_index( ctx->repair, repair_req->slot, repair_req->shred_index );
+        break;
+      }
+      case FD_REPAIR_REQ_TYPE_NEED_HIGHEST_WINDOW_INDEX: {
+        rc = fd_repair_need_highest_window_index( ctx->repair, repair_req->slot, repair_req->shred_index );
+        break;
+      }
+      case FD_REPAIR_REQ_TYPE_NEED_ORPHAN: {
+        rc = fd_repair_need_orphan( ctx->repair, repair_req->slot );
+        break;
+      }
+    }
+    
+    if( rc < 0 ) {
+      FD_LOG_WARNING(( "failed to issue repair request" ));
+    }
+  }
+
+}
+
+static inline void
+handle_new_stake_weights( fd_repair_tile_ctx_t * ctx,
+                          uchar const    * buf ) {
+  ulong const * header = (ulong const *)fd_type_pun_const( buf );
+
+  ulong stakes_cnt = header[ 1 ]; 
+
+  if( stakes_cnt >= MAX_REPAIR_PEERS ) {
+    FD_LOG_ERR(( "Cluster nodes had %lu stake weights, which was more than the max of %lu", stakes_cnt, MAX_REPAIR_PEERS ));
+  }
+
+  fd_stake_weight_t const * in_stake_weights = fd_type_pun_const( header+4UL );
+  fd_repair_set_stake_weights( ctx->repair, in_stake_weights, stakes_cnt );
+}
+
+
+static void 
+repair_send_packet( uchar const * msg, 
+                    size_t msglen, 
+                    fd_gossip_peer_addr_t const * addr, 
+                    void * arg ) {
+  ulong tsorig = fd_frag_meta_ts_comp( fd_tickcount() );
+  send_packet( arg, addr->addr, fd_ushort_bswap( addr->port ), msg, msglen, tsorig );
+}
+
+static void
+repair_shred_deliver( fd_shred_t const * shred,
+                      ulong shred_sz, 
+                      fd_repair_peer_addr_t const * from FD_PARAM_UNUSED, 
+                      fd_pubkey_t const * id FD_PARAM_UNUSED, 
+                      void * arg ) {
+  fd_repair_tile_ctx_t * ctx = (fd_repair_tile_ctx_t *)arg;
+  ulong tsorig = fd_frag_meta_ts_comp( fd_tickcount() );
+  
+  fd_shred_t * out_shred = fd_chunk_to_laddr( ctx->store_out_mem, ctx->store_out_chunk );
+  fd_memcpy( out_shred, shred, shred_sz );
+  
+  ulong tspub = fd_frag_meta_ts_comp( fd_tickcount() );
+  fd_mcache_publish( ctx->store_out_mcache, ctx->store_out_depth, ctx->store_out_seq, 1, ctx->store_out_chunk,
+    shred_sz, 0UL, tsorig, tspub );
+  ctx->store_out_seq   = fd_seq_inc( ctx->store_out_seq, 1UL );
+  ctx->store_out_chunk = fd_dcache_compact_next( ctx->store_out_chunk, shred_sz, ctx->store_out_chunk0, ctx->store_out_wmark );}
+
+static void
+repair_shred_deliver_fail( fd_pubkey_t const * id FD_PARAM_UNUSED, 
```

Comment:
Why can this happen? What does it mean for a shred to fail to be delivered?

Reply:
I didn't write any of the repair code but afaict, this is a callback passed into the library code so it just needs to exist. From a quick read through, does not look like we can hit a case which triggers this from the repair tile, but just having a log for completeness makes sense. This becomes a more important callback for replay

Reply:
As for when this happens, the condition is when the "needed_table" is full (i.e. too many reqs)